operator/internal/controller/console/controller.go

```diff
@@ -0,0 +1,331 @@
+	// value.
+	rng *rand.Rand
+}
+
+func (c *Controller) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	// If rng is not set for testing, create and seed a new one.
+	if c.rng == nil {
+		// TODO: Weak RNG is probably acceptable here but best to doublecheck
+		c.rng = rand.New(rand.NewSource(time.Now().UnixMicro())) //nolint:gosec
+	}
+
+	// Setup a field index for clusterRef so we can trigger updates to Consoles when Redpandas are changed.
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &redpandav1alpha2.Console{}, clusterReferenceIndex, func(o client.Object) []string {
+		console := o.(*redpandav1alpha2.Console)
+		if console.Spec.ClusterSource == nil || console.Spec.ClusterSource.ClusterRef == nil {
+			return nil
+		}
+		return []string{
+			kube.ObjectKey{Namespace: console.Namespace, Name: console.Spec.ClusterSource.ClusterRef.Name}.String(),
+		}
+	}); err != nil {
+		return errors.WithStack(err)
+	}
+
+	builder := ctrl.NewControllerManagedBy(mgr)
+
+	// NB: As of writing, all console types are namespace scoped.
+	for _, t := range console.Types() {
+		builder = builder.Owns(t)
+	}
+
+	return builder.
+		For(&redpandav1alpha2.Console{}).
+		// Configure a watch on redpandas using controller-runtime's indexing.
+		// If a redpanda is updated, any console's referring to it will be
+		// re-reconciled.
+		Watches(&redpandav1alpha2.Redpanda{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
+			key := kube.AsKey(o)
+
+			list, err := kube.List[redpandav1alpha2.ConsoleList](ctx, c.Ctl, kube.NamespaceAll, &client.ListOptions{
+				FieldSelector: fields.OneTermEqualSelector(clusterReferenceIndex, key.String()),
+			})
+			if err != nil {
+				mgr.GetLogger().V(1).Info(fmt.Sprintf("possibly skipping %T reconciliation due to failure to fetch %s associated with cluster", &redpandav1alpha2.Console{}, clusterReferenceIndex), "error", err)
+				return nil
+			}
+
+			reqs := make([]reconcile.Request, len(list.Items))
+			for i, console := range list.Items {
+				reqs[i] = reconcile.Request{
+					NamespacedName: kube.AsKey(&console),
+				}
+			}
+			return reqs
+		})).
+		Complete(c)
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cr, err := kube.Get[redpandav1alpha2.Console](ctx, c.Ctl, req.NamespacedName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if err := c.maybeSetJWTToken(ctx, cr); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	gvk, err := kube.GVKFor(c.Ctl.Scheme(), cr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	syncer := kube.Syncer{
+		Ctl:             c.Ctl,
+		Namespace:       cr.Namespace,
+		Renderer:        c.rendererFor(cr),
+		Owner:           *metav1.NewControllerRef(cr, gvk),
+		OwnershipLabels: c.ownershipLabelsFor(cr),
+	}
+
+	objs, err := syncer.Sync(ctx)
```

Comment:
Looks like we're relying heavily on the GC'ing behavior of ownership here as we have no finalizer cleanup routine? That could potentially orphan any cluster-scoped objects which we eventually add as you can't set a real ownerRef on those (the comment above about all of the objects currently being namespace-scoped aside). Just for future extensibility I'd probably prefer we do the whole finalizer + cleanup routine just so we don't forget if we need to add something cluster scoped.

Reply:
Done!